### PR TITLE
upgraded python to 3.9 for bamboo image

### PR DESCRIPTION
**Pull request by rizbihassan**
## Summary of Changes

Addresses python version upgrade to 3.9 in bamboo image

## Changes

* updated base_images/Dockerfile.local
* updated base_images/Dockerfile.bamboo

## Type of change

Please delete options that are not relevant.

- [ x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

tested locally

## Checklist:

- [x ] My code follows the style guidelines of this project
- [x ] I have performed a self-review of my own code
- [ x] I have commented my code, particularly in hard-to-understand areas
- [ x] My code has passed security scanning
    - [x ] Snyk
    - [x ] git-secrets

### CODE REVIEW

File: base_images/Dockerfile.bamboo [hunk 1]
```diff
@@ -14,11 +14,21 @@ curl "https://s3.amazonaws.com/session-manager-downloads/plugin/latest/linux_64b
...
-yum install -y jq gcc git make openssl wget zip unzip tar python3-devel nodejs yarn python3-pip awscli session-manager-plugin.rpm procps parallel && \
+yum install -y jq gcc git make openssl wget zip unzip tar nodejs yarn python3-pip awscli session-manager-plugin.rpm procps parallel && \
 pip3 install -q --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org pydoc-markdown bandit safety flake8 black isort
 
+
+# install Python 3.9
+RUN \
+yum install -y openssl-devel bzip2-devel libffi-devel && \
+wget https://www.python.org/ftp/python/3.9.10/Python-3.9.10.tgz && \
+tar xvf Python-3.9.10.tgz && \
```
Comment (AndrewDorn) [inline base_images/Dockerfile.bamboo:27]: In both these files, let's pull `3.9.10` into a variable to make updating easier.

Comment (AndrewDorn) [inline base_images/Dockerfile.bamboo:27]: We'll need to remember to check for updates on occasion. Not really related to this change, as clearly our library went out-of-date when we weren't checking anyway.

Comment (rizbihassan) [reply base_images/Dockerfile.bamboo]: updated

**Review by AndrewDorn — APPROVED** [review]

File: base_images/Dockerfile.bamboo [hunk 2]
```diff
@@ -14,11 +15,22 @@ curl "https://s3.amazonaws.com/session-manager-downloads/plugin/latest/linux_64b
 amazon-linux-extras install epel -y && \
 yum update -y
 
-# install Python 3, pip, CLI utilities, session manager and other tools
+# install pip, CLI utilities, session manager and other tools
 RUN \
-yum install -y jq gcc git make openssl wget zip unzip tar python3-devel nodejs yarn python3-pip awscli session-manager-plugin.rpm procps parallel && \
+yum install -y jq gcc git make openssl wget zip unzip tar nodejs yarn python3-pip awscli session-manager-plugin.rpm procps parallel && \
```
Comment (bhazuka) [inline base_images/Dockerfile.bamboo:20]: Remove `python3-pip` from the list and add the development libraries needed to install python 3.9 `openssl-devel bzip2-devel libffi-devel ncurses-devel sqlite-devel readline-devel uuid-devel libuuid-devel gdbm-devel xz-devel`

I would also install development tools as well to get additional options in Python
`yum groupinstall "Development Tools && \`

Note you will need to build python before running the pip command on line 21. I will add additional notes in your install python block. I would move the entire Python build into this block.

File: base_images/Dockerfile.bamboo [hunk 3]
```diff
@@ -5,6 +5,7 @@ LABEL maintainer="Rizbi Hassan" \
 
 ENV NODE_VERSION="16.x"
 ENV TERRAFORM_VERSION "0.13.6"
+ENV PYTHON_VERSION "3.9.10"
```
Comment (bhazuka) [inline base_images/Dockerfile.bamboo:8]: I would pull Python version 3.9.16 as it is the lateset 3.9 series with bug fixes.

File: base_images/Dockerfile.bamboo [hunk 4]
```diff
@@ -14,11 +15,22 @@ curl "https://s3.amazonaws.com/session-manager-downloads/plugin/latest/linux_64b
...
 pip3 install -q --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org pydoc-markdown bandit safety flake8 black isort
 
+
+# install Python 3.9
+RUN \
+yum install -y openssl-devel bzip2-devel libffi-devel && \
+wget https://www.python.org/ftp/python/${PYTHON_VERSION}/Python-${PYTHON_VERSION}.tgz && \
+tar xvf Python-${PYTHON_VERSION}.tgz && \
+cd Python-*/ && \
+./configure --enable-optimizations && \
```
Comment (bhazuka) [inline base_images/Dockerfile.bamboo:30]: Add `--enable-loadable-sqlite-extensions --enable-ipv6` to the configuration options.

File: base_images/Dockerfile.bamboo [hunk 5]
```diff
@@ -14,11 +15,22 @@ curl "https://s3.amazonaws.com/session-manager-downloads/plugin/latest/linux_64b
...
+make altinstall && \
```
Comment (bhazuka) [inline base_images/Dockerfile.bamboo:31]: After the altinstall run the following commands as to not break our current scripts.
```
ln -sf /usr/local/bin/python3.9 /usr/local/bin/python3 && \
ln -sf /url/local/bin/pip3.9 /usr/local/bin/pip3
```

File: base_images/Dockerfile.bamboo [hunk 6]
```diff
@@ -14,11 +15,22 @@ curl "https://s3.amazonaws.com/session-manager-downloads/plugin/latest/linux_64b
...
+rm -f Python-${PYTHON_VERSION}.tgz
```
Comment (bhazuka) [inline base_images/Dockerfile.bamboo:32]: Remove your build directory as well

```
rm -f Python-${PYTHON_VERSION}.tgz && \
rm -rf Python-${PYTHON_VERSION}
```

File: bin/common/venv_management.sh [hunk 7]
```diff
@@ -10,7 +10,7 @@ function create_and_activate_venv () {
     find . -type d -name "__pycache__" -exec rm -rf {} +
   fi
   
-  run_and_check_returncode "python3 -m venv venv"
+  run_and_check_returncode "python3.9 -m venv venv"
```
Comment (bhazuka) [inline bin/common/venv_management.sh:13]: Revert back to python3 to keep it generalized for future python upgrades

File: base_images/Dockerfile.local [hunk 8]
```diff
@@ -5,6 +5,7 @@ LABEL maintainer="Rizbi Hassan" \
 
 ENV NODE_VERSION="16.x"
 ENV TERRAFORM_VERSION "0.13.6"
+ENV PYTHON_VERSION "3.9.10"
```
Comment (bhazuka) [inline base_images/Dockerfile.local:8]: See notes on changes made to the bamboo Dockerfile and replicate here.

**Review by bhazuka — CHANGES_REQUESTED** [review]
Overall looks good. Some rearranging to move python into the install layer. Some additional libraries and build options to give Python more functionality and capability. Additional cleanup measures.

File: base_images/Dockerfile.bamboo (hunk 2 above)
Comment (rizbihassan) [reply base_images/Dockerfile.bamboo:20]: updated as stated above. For some reason, I am getting error while running 
`pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org pydoc-markdown`

Got same error as well with this
`pip3.9 install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org pydoc-markdown`


Error:
#6 323.4 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#6 323.5 /bin/sh: pip: command not found

File: base_images/Dockerfile.bamboo (hunk 3 above)
Comment (rizbihassan) [reply base_images/Dockerfile.bamboo:8]: updated

File: base_images/Dockerfile.bamboo (hunk 4 above)
Comment (rizbihassan) [reply base_images/Dockerfile.bamboo:30]: added

File: base_images/Dockerfile.bamboo (hunk 5 above)
Comment (rizbihassan) [reply base_images/Dockerfile.bamboo:31]: added

File: base_images/Dockerfile.bamboo (hunk 6 above)
Comment (rizbihassan) [reply base_images/Dockerfile.bamboo:32]: removed

File: base_images/Dockerfile.local (hunk 8 above)
Comment (rizbihassan) [reply base_images/Dockerfile.local:8]: updated the changes here as well

**Comment by rizbihassan** [comment]
tested running unit tests in bamboo by commneting out safety and it now runs in python 3.9.16 without any error
https://ci.earthdata.nasa.gov/browse/ORCA-OI291-JOB1-2/log

File: base_images/Dockerfile.bamboo (hunk 2 above)
Comment (rizbihassan) [reply base_images/Dockerfile.bamboo]: looks like we don't even need this pip install comment for running unit tests in bamboo.
All lambas bin/run_tests.sh already have this line of code

**Comment by rizbihassan** [comment]
after testing in bamboo, unit tests are successful using python 3.9. Note that I did comment out safety to avoid its interference with the tests
https://ci.earthdata.nasa.gov/browse/ORCA-OI291-JOB1-5

**Review by bhazuka — APPROVED** [review]

**Pull request merged**